src/matrix/room/timeline/Timeline.js

```diff
@@ -236,8 +247,50 @@ export class Timeline {
     /** @package */
     addEntries(newEntries) {
         this._addLocalRelationsToNewRemoteEntries(newEntries);
+        this._updateFetchedEntries(newEntries);
+        this._loadRelatedEvents(newEntries);
         this._remoteEntries.setManySorted(newEntries);
     }
+
+    _updateFetchedEntries(entries) {
+        for (const entry of entries) {
+            const relatedEntry = this._fetchedEventEntries.find(e => e.id === entry.relatedEventId);
+            if (relatedEntry?.addLocalRelation(entry)) {
+                relatedEntry.dependents.forEach(e => this._findAndUpdateRelatedEntry(null, e.id, () => true));
+            }
+        }
+    }
+
+    async _loadRelatedEvents(entries) {
+        const filteredEntries = entries.filter(e => !!e.contextEventId);
+        for (const entry of filteredEntries) {
+            const id = entry.contextEventId;
+            let needToTrack = false;
+            let contextEvent;
+            // find in remote events
+            contextEvent = this.getByEventId(id);
+            // find in storage
+            if (!contextEvent) {
+                needToTrack = true;
+                contextEvent = await this._fetchEventFromStorage(id);
+            }
+            // fetch from hs
+            if (!contextEvent) {
+                contextEvent = await this._fetchEventFromHomeserver(id);
+            }
+            if (contextEvent) {
+                if (needToTrack) {
+                    // this entry was created from storage/hs, so it's not tracked by remoteEntries
+                    // we track them here so that we can update reply preview of dependents on redaction
+                    this._fetchedEventEntries.push(contextEvent);
```

Comment:
If this only needs to happen if the entry comes from storage, can we move this below `await this._fetchEventFromStorage(id);` on line 275? Looks like we wouldn't need the `needToTrack` flag then? If you can replace the flag with control flow, that might be more readable.

OTOH, I wonder why we wouldn't need to track if the event came from the homeserver? Isn't it possible an event we don't have in storage gets redacted?